src/three-components/PMREMGenerator.ts

```diff
@@ -0,0 +1,367 @@
+    this.renderer.toneMapping = toneMapping;
+    this.renderer.toneMappingExposure = toneMappingExposure;
+    this.renderer.gammaInput = gammaInput;
+    this.renderer.gammaOutput = gammaOutput;
+  }
+
+  packMipmaps() {
+    this.objects.forEach((object) => {
+      this.packingScene.add(object);
+    });
+
+    const gammaInput = this.renderer.gammaInput;
+    const gammaOutput = this.renderer.gammaOutput;
+    const toneMapping = this.renderer.toneMapping;
+    const toneMappingExposure = this.renderer.toneMappingExposure;
+    const currentRenderTarget = this.renderer.getRenderTarget();
+
+    this.renderer.gammaInput = false;
+    this.renderer.gammaOutput = false;
+    this.renderer.toneMapping = LinearToneMapping;
+    this.renderer.toneMappingExposure = 1.0;
+    this.renderer.setRenderTarget(this.cubeUVRenderTarget);
+    this.renderer.render(this.packingScene, this.flatCamera);
+
+    this.renderer.setRenderTarget(currentRenderTarget);
+    this.renderer.toneMapping = toneMapping;
+    this.renderer.toneMappingExposure = toneMappingExposure;
+    this.renderer.gammaInput = gammaInput;
+    this.renderer.gammaOutput = gammaOutput;
+
+    this.objects.forEach((object) => {
+      this.packingScene.remove(object);
+    });
+    this.cubeLods.forEach((target) => {
+      target.dispose();
+    });
+    this.cubeLods = [];
+  }
+
+  getMipmapShader() {
+    var shaderMaterial = new ShaderMaterial({
+
+      uniforms: {'texelSize': {value: 0.5}, 'envMap': {value: null}},
+
+      vertexShader: `
+varying vec2 vUv;
+varying vec3 vPosition;
+void main() {
+    vUv = uv;
+    vPosition = position;
+    gl_Position = projectionMatrix * modelViewMatrix * vec4( position, 1.0 );
+}
+`,
+
+      fragmentShader: `
+varying vec2 vUv;
+varying vec3 vPosition;
+uniform float texelSize;
+uniform samplerCube envMap;
+int getFace(vec3 direction) {
+    vec3 absDirection = abs(direction);
+    int face = -1;
+    if (absDirection.x > absDirection.z) {
+      if (absDirection.x > absDirection.y)
+        face = direction.x > 0.0 ? 0 : 3;
+      else
+        face = direction.y > 0.0 ? 1 : 4;
+    } else {
+      if (absDirection.z > absDirection.y)
+        face = direction.z > 0.0 ? 2 : 5;
+      else
+        face = direction.y > 0.0 ? 1 : 4;
+    }
+    return face;
+}
+vec3 getDirection(vec2 uv, int face) {
+    uv = 2.0 * uv - 1.0;
+    vec3 direction;
+    if (face == 0) {
+      direction = vec3(1.0, uv.y, -uv.x);
+    } else if (face == 1) {
+      direction = vec3(uv.x, 1.0, -uv.y);
+    } else if (face == 2) {
+      direction = vec3(uv, 1.0);
+    } else if (face == 3) {
+      direction = vec3(-1.0, uv.y, uv.x);
+    } else if (face == 4) {
+      direction = vec3(uv.x, -1.0, uv.y);
+    } else {
+      direction = vec3(-uv.x,uv.y, -1.0);
+    }
+    return direction;
+}
+void main() {
+  int face = getFace(vPosition);
+  vec2 uv = vUv - 0.5 * texelSize;
+  vec3 texelDir = getDirection(uv, face);
```

Comment:
My general attitude about abbreviation is that we should avoid it. I personally find it less readable. But, for GLSL it's a different language so I suppose we could afford to have a different convention (I understand that culturally speaking, folks authoring shaders enjoy their abbreviations, and Three.js is no exception here). Open to suggestions here if you feel strongly about it one way or another, but if you don't then let's carry over the JS convention.

Reply:
Well, my intention is to upstream this to Three.js (assuming it works well), so I think it's more important to match their coding conventions.